rust/agama-dbus-server/src/network/nm/dbus.rs

```diff
@@ -547,9 +583,64 @@ fn wireless_config_from_dbus(conn: &OwnedNestedHash) -> Option<WirelessConfig> {
         ..Default::default()
     };
 
+    if let Some(band) = wireless.get("band") {
+        wireless_config.band = Some(band.downcast_ref::<str>()?.try_into().ok()?)
+    }
+    if let Some(channel) = wireless.get("channel") {
+        wireless_config.channel = Some(*channel.downcast_ref()?);
+    }
+    if let Some(bssid) = wireless.get("bssid") {
+        let bssid: &zvariant::Array = bssid.downcast_ref()?;
+        let bssid: Vec<u8> = bssid
+            .get()
+            .iter()
+            .map(|u| *u.downcast_ref::<u8>().unwrap())
+            .collect();
+        wireless_config.bssid = Some(MacAddr6::new(
+            *bssid.first()?,
+            *bssid.get(1)?,
+            *bssid.get(2)?,
+            *bssid.get(3)?,
+            *bssid.get(4)?,
+            *bssid.get(5)?,
+        ));
+    }
+
     if let Some(security) = conn.get(WIRELESS_SECURITY_KEY) {
         let key_mgmt: &str = security.get("key-mgmt")?.downcast_ref()?;
         wireless_config.security = NmKeyManagement(key_mgmt.to_string()).try_into().ok()?;
+
+        let wep_key_type = if let Some(wep_key_type) = security.get("wep-key-type") {
+            let wep_key_type: u32 = *wep_key_type.downcast_ref()?;
+            match wep_key_type {
+                // 0 shouldn't appear because it is treated as empty but just in case
+                0 => WepKeyType::Unknown,
+                1 => WepKeyType::Key,
+                2 => WepKeyType::Passphrase,
+                _ => {
+                    log::error!("\"wep-key-type\" from NetworkManager not valid");
+                    WepKeyType::default()
+                }
+            }
+        } else {
+            WepKeyType::default()
+        };
+        let auth_alg = if let Some(auth_alg) = security.get("auth-alg") {
```

Comment:
I usually find using combinators more readable, but I guess it is a matter of taste (I am used to them in other languages):

```rust
let auth_alg = security
    .get("auth-alg")
    .and_then(|alg| WepAuthAlg::try_from(alg.downcast_ref()?).ok())
    .unwrap_or_default();
```

And the same for the index:

```rust
let wep_key_index = security
    .get("wep-tx-keyidx")
    .and_then(|idx| idx.downcast_ref::<u32>().cloned())
    .unwrap_or_default();
```

And I guess you would follow a similar approach above if you move the complex part (the match) to a separate function.

But feel free to ignore these suggestions.